Add support of Cryptomatte AOV

usakhelo opened the pull request:
Hi guys!

This is an attempt to add Cryptomatte AOV. It's done in the same way as denoiser, so in order to save cryptomatte AOV you will need to enable it on beauty output and save main output after rendering. It has two modes - object names and material names. It works fine on my simple test scene where I used depth of field and opacity to test matte selection. I tested it on Nuke and Fusion.

Thanks for review!

est77 commented on CMakeLists.txt:
```diff
@@ -283,6 +283,7 @@ endif ()
 include_directories (
     src/thirdparty/bcd
     src/thirdparty/lz4
+    src/thirdparty/murmurhash3
```
We already have murmurhash in foundation/utility, wrapped in a class. It looks like the same version.
Use that one, modifying it if needed, instead of adding another copy of the code.

usakhelo replied:
Their specification requires "32 bit variant of MurmurHash3 (Also known as MurmurHash3_x86_32)"
The foundation version is a x64_128 bit version. I tried to use that hash truncating it to 32 bits but it didn't work. Is there any way I can get the same 32 bit value from 128bits?

usakhelo replied:
Maybe I should add x86_32 version to an existing `MurmurHash` class?

est77 replied:
I see. I think that we can keep it like it is, as an implementation detail of cryptomatte.

usakhelo replied:
Maybe I should remove all other implementations from external code and leave only x86_32bit?

est77 commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,736 @@
+        for (size_t rx = 0, e = impl->m_image->properties().m_canvas_width; rx < e; ++rx)
+        {
+                vector<float> pixel_values;
+                for (size_t i = 0; i < num_channels; ++i)
```
Move this out of the loop. Use the vector constructor that creates N elements with a value.

est77 commented on src/appleseed/renderer/modeling/frame/frame.cpp:
```diff
@@ -952,6 +970,25 @@ void Frame::extract_parameters()
+
+    // Retrieve cryptomatte parameters.
+    {
+        const string cryptomatte_mode = m_params.get_optional<string>("cryptomatte", "off");
```
I think we can remove this now

est77 commented on src/appleseed/renderer/modeling/frame/frame.h:
```diff
@@ -170,6 +170,12 @@ class APPLESEED_DLLSYMBOL Frame
         const size_t                thread_count,
         foundation::IAbortSwitch*   abort_switch) const;
 
+    enum class CryptomatteMode
```
Remove this enum too.

est77 commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,752 @@
+
+DictionaryArray CryptomatteAOVFactory::get_input_metadata() const
+{
+    DictionaryArray metadata;
```
We can add here the option to choose between object names or material names.
We don't have a way to edit them in studio but we'll have it.

Another option would be to have 2 Cryptomatte AOVs. One for object IDs and one for material IDs.
This option might be easier to support in plugins. I'll check that.

usakhelo replied:
I thought about two AOVs too, it will require another line `register_factory(auto_release_ptr<FactoryType>(new CryptomatteAOVFactory()));`
But mode parameter can passed to Factory class maybe?

usakhelo commented:
Hi @est77 , I've fixed comments and changed Factory so that it can create two different AOVs. Tested in 3ds max and it works. This branch requires rebase I guess.

dictoon commented on src/appleseed/renderer/modeling/aov/aovfactoryregistrar.cpp:
```diff
@@ -32,6 +32,7 @@
 // appleseed.renderer headers.
 #include "renderer/modeling/aov/aovtraits.h"
 #include "renderer/modeling/aov/albedoaov.h"
+#include "renderer/modeling/aov/cryptomatte.h"
```
The files should be called `cryptomatteaov.{h,cpp}`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.h:
```diff
@@ -0,0 +1,133 @@
+#include "foundation/utility/autoreleaseptr.h"
+
+// BCD headers.
+#include "bcd/DeepImage.h"
```
This doesn't look needed?

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.h:
```diff
@@ -0,0 +1,133 @@
+  private:
+    friend class CryptomatteAOVFactory;
+
+    struct                      Impl;
```
We only require that data members are aligned with other data members. Since there are no others here, you don't need all these spaces.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.h:
```diff
@@ -0,0 +1,133 @@
+        const size_t    canvas_height,
+        const size_t    tile_width,
+        const size_t    tile_height,
+        ImageStack&     aov_images) override;
```
`ImageStack` needs to be forward-declared.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.h:
```diff
@@ -0,0 +1,133 @@
+        MaterialNames,
+    };
+
+    CryptomatteAOV(const ParamArray& params);
```
Mark constructor as `explicit`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.h:
```diff
@@ -0,0 +1,133 @@
+
+// Forward declarations.
+namespace foundation    { class ImageAttributes; }
+namespace renderer      { class AOV; }
```
Not required since you include `aov.h`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.h:
```diff
@@ -0,0 +1,133 @@
+//
+
+class APPLESEED_DLLSYMBOL CryptomatteAOVFactory
+    : public IAOVFactory
```
Indent with two spaces.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.h:
```diff
@@ -0,0 +1,133 @@
+
+
+//
+// A factory for cryptomatte AOVs.
```
*Cryptomatte*

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.h:
```diff
@@ -0,0 +1,133 @@
+      const char* get_model() const override;
+
+      // Return metadata for this AOV model.
+      foundation::Dictionary get_model_metadata() const override;
```
Foward declare `foundation::Dictonary`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.h:
```diff
@@ -0,0 +1,133 @@
+      foundation::Dictionary get_model_metadata() const override;
+
+      // Return metadata for the inputs of this AOV model.
+      foundation::DictionaryArray get_input_metadata() const override;
```
Foward declare `foundation::DictonaryArray`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+#include "renderer/kernel/shading/shadingcomponents.h"
+#include "renderer/kernel/shading/shadingpoint.h"
+#include "renderer/kernel/shading/shadingresult.h"
+#include "renderer/modeling/aov/aov.h"
```
Not required since already included in the header.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+
+// Standard headers.
+#include <cmath>
+#include <cstddef>
```
Not required for the same reason.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+namespace
+{
+    class MultiChannelExrFileWriter
+        : public IImageFileWriter
```
Indent with two spaces.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+    {
+      public:
+        explicit MultiChannelExrFileWriter(const char* filename, const ICanvas* image)
+            : m_canvas(image)
```
Indent with two spaces.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+                throw ExceptionIOError(msg.c_str());
+            }
+
+            // m_writer->write_image(OIIO::TypeDesc::FLOAT, m_canvas);
```
Remove dead code?

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+
+                if (attr_name == "software")
+                    m_spec.attribute("Software", attr_value.c_str());
+
```
Remove blank lines between `if` / `else if` blocks.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+            set_image_spec();
+        }
+
+        ~MultiChannelExrFileWriter()
```
Mark all base class methods you override with `override`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+
+        void set_image_attributes(const ImageAttributes& image_attributes)
+        {
+            for (const_each<ImageAttributes> i = image_attributes; i; ++i)
```
Prefer range-based `for` loops in new code.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+
+        void set_image_channels(
+            const size_t            channel_count,
+            const vector<string>&   channel_names)
```
Include `<vector>`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+
+                    // Write the tile into the file.
+                    if (!m_writer->write_tile(
+                        static_cast<int>(tile_offset_x),
```
Indent method arguments once more.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+            m_writer = OIIO::ImageOutput::create(m_filename);
+            if (m_writer == nullptr)
+            {
+                const string msg = OIIO::geterror();
```
Include `<string>`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+        const char*         m_filename;
+    };
+
+    float hash_to_float(uint32_t hash) {
```
Move brace to its own line.
Add reference to where you took this code from.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+        if (exponent == 0 || exponent == 255)
+            hash ^= 1 << 23; // toggle bit
+        float f;
+        memcpy(&f, &hash, 4);
```
Include `<cstring>`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+namespace
+{
+
+    typedef map<uint32_t, string> NameMap;
```
Include `<map>`.
Use `uint32` from `foundation` and include `foundation/platform/types.h`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+
+    typedef map<uint32_t, string> NameMap;
+    typedef map<uint32_t, float> WeightMap;
+
```
Two blank lines before block comment.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+            map<pair<int, int>, NameMap>&       tile_name_map,
+            int                                 num_layers,
+            CryptomatteAOV::CryptomatteType     layer_type)
+            : m_aov_image(aov_image)
```
Indent only with two spaces.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+      public:
+        CryptomatteAOVAccumulator(
+            Image&                              aov_image,
+            vector<WeightMap>&                  pixel_samples,
```
A vector of maps? Is that intended?

If it is, how many maps will we have? One per pixel in the tile?

How many items will each map typically hold? Could a lighter (probably home-grown) data structure be used instead of `std::map<>`?

usakhelo replied:
Hi @dictoon ,
Yes it's a map per pixel, it's variable size and holds up to the amount of objects that share the pixel. Typically up to 3-4 in test scenes. I'll check if there is anything lighter in foundation.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+                {
+                    const WeightMap& weight_map = m_pixel_samples[ry * m_frame_width + rx];
+
+                    if (weight_map.size() > 0)
```
`!weight_map.empty()`

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+                        if (total_weight == 0.0f)
+                            total_weight = 1.0f;
+
+                        vector<pair<float, uint32_t>> ranked_vector;
```
Reserve memory in advance since you know exactly how many items the vector will hold.

Even better would be to have a single vector allocated outside of the loop, and cleared (using `foundation::clear_keep_memory()`) every time. You wouldn't need to reserve memory anymore and you would save a lot of memory allocations and deallocations.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+                        vector<pair<float, uint32_t>> ranked_vector;
+                        for (const auto& item : weight_map)
+                            ranked_vector.push_back(make_pair(item.second, item.first));
+                        sort(ranked_vector.begin(), ranked_vector.end(),
```
Include `<algorithm>`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+                        for (const auto& item : weight_map)
+                            ranked_vector.push_back(make_pair(item.second, item.first));
+                        sort(ranked_vector.begin(), ranked_vector.end(),
+                            [](pair<float, uint32_t> const& a, pair<float, uint32_t> const& b) { return a.first > b.first; });
```
Include `<utility>`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+
+                        vector<float> pixel_values;
+                        pixel_values.push_back(m3hash == 0 ? 0.0f : hash_to_float(m3hash));
+                        pixel_values.push_back(m3hash == 0 ? 0.0f : ((float)((m3hash << 8)) / (float)UINT32_MAX));
```
Use `static_cast<>()` instead of C-style casts.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+                        uint32_t m3hash = ranked_vector[0].second;
+
+                        vector<float> pixel_values;
+                        pixel_values.push_back(m3hash == 0 ? 0.0f : hash_to_float(m3hash));
```
Test `m3hash` just once.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+                        pixel_values.push_back(m3hash == 0 ? 0.0f : ((float)((m3hash << 8)) / (float)UINT32_MAX));
+                        pixel_values.push_back(m3hash == 0 ? 0.0f : ((float)((m3hash << 16)) / (float)UINT32_MAX));
+
+                        // Remove background contribution
```
End full line comments with a period.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+                        if (ranked_vector.size() > 1 && m3hash == 0)
+                            ranked_vector.erase(ranked_vector.begin());
+
+                        for (size_t i = 0, e = ranked_vector.size(); i < e && i < m_num_layers; ++i)
```
Set `e` to `min(ranked_vector.size(), m_num_layers)`?

usakhelo replied:
Thanks for this tip.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+                        }
+
+                        const int num_channels = (m_num_layers * 2) + 3;
+                        for (size_t i = ranked_vector.size(); i < num_channels; ++i)
```
You're mixing `int` and `size_t` here. Use `size_t`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+
+
+//
+// Cryptomatte AOV class implementation.
```
The class is called `CryptomatteAOV`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+  : AOV("cryptomatte", params)
+  , impl(new Impl())
+{
+    const string cryptomatte_type = params.get_optional<string>("cryptomatte_type", "object_names");
```
You can pass a list of valid values to `get_optional()`, that way an error message will be emitted if the value is not valid.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+    const int h = static_cast<int>(canvas_height);
+
+    // Create underlying images.
+    const int num_channels = (impl->m_num_layers * 2) + 3;
```
`size_t`

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+    const size_t    tile_height,
+    ImageStack&     aov_images)
+{
+    const int w = static_cast<int>(canvas_width);
```
Why the casts? Those variables don't seem to be useful, just use `canvas_width` and `canvas_height`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+
+void CryptomatteAOV::clear_image()
+{
+    const int num_channels = (impl->m_num_layers * 2) + 3;
```
`size_t`

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+{
+    const int num_channels = (impl->m_num_layers * 2) + 3;
+    vector<float> pixel_values(num_channels, 0.0f);
+    for (size_t ry = 0, e = impl->m_image->properties().m_canvas_height; ry < e; ++ry)
```
Call `impl->m_image->properties()` once.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+    vector<float> pixel_values(num_channels, 0.0f);
+    for (size_t ry = 0, e = impl->m_image->properties().m_canvas_height; ry < e; ++ry)
+    {
+        for (size_t rx = 0, e = impl->m_image->properties().m_canvas_width; rx < e; ++rx)
```
You're using `e` in both loops.

usakhelo replied:
Terrible. How it ever worked?!

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+        break;
+
+      default:
+        break;
```
Use `assert_otherwise` from `foundation/utility/otherwise.h`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+    const string exr_file_name = base_file_name + ".cryptomatte" + extension;
+    const string exr_file_path = (directory / exr_file_name).string();
+
+    string layer_name = "AppleseedObjectName";
```
I don't think you need to initialize the variable here?

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+    NameMap name_map;
+    for (auto& pair : impl->m_tile_name_map)
+    {
+        for (auto& hash : pair.second)
```
`const`?

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+    channel_names.push_back("B");
+
+    int num_exr_layers = truncate<int>(fast_ceil(impl->m_num_layers / 2.0));
+    for(size_t i = 0; i < num_exr_layers; ++i)
```
Space after `for`.
Mixing signed/unsigned here. Use `size_t`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+    int num_exr_layers = truncate<int>(fast_ceil(impl->m_num_layers / 2.0));
+    for(size_t i = 0; i < num_exr_layers; ++i)
+    {
+        string layer_number = get_numbered_string("##", i);
```
Make const.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,784 @@
+//
+
+CryptomatteAOVFactory::CryptomatteAOVFactory(const CryptomatteAOV::CryptomatteType aov_type)
+    : m_aov_type(aov_type)
```
Indent with two spaces.

usakhelo commented:
Hi @dictoon , I've fixed the comments and replaced per-pixel map with in-house simple object. Maybe it is super naive, I'm not sure.
Also I'm not sure if I should do rebase, because CI complains.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -402,7 +527,7 @@ namespace
             ShadingResult&              shading_result) override
         {
-            uint32_t m3hash = 0;
+            uint32 m3hash = 0;
```
I can't comment further down in this method, but can you mark as `const` all variables that can be const?

dictoon replied:
You can also precompute `1.0 / m_rcp_canvas_width` etc.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,908 @@
+    {
+        for (size_t rx = 0, ex = image_props.m_canvas_width; rx < ex; ++rx)
+        {
+                impl->m_image->set_pixel(rx, ry, pixel_values.data());
```
Fix indentation.

dictoon replied:
That's going to be very slow because for each pixel `ICanvas::set_pixel()` needs to find the associated tile.

Here's a faster implementation (untested, please check):

```cpp
for (size_t ty = 0, ey = image_props.m_tile_count_y; ty < ey; ++ty)
{
    for (size_t tx = 0, ex = image_props.m_tile_count_x; tx < ex; ++tx)
    {
        Tile& tile = impl->m_image->tile(tx, ty);
        memset(tile.get_storage(), 0, tile.get_size());
    }
}
```

That works because `0` is the bit pattern for `0.0f` (how convenient!).

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,908 @@
+
+void CryptomatteAOV::clear_image()
+{
+    const size_t num_channels = (impl->m_num_layers * 2) + 3;
```
The expression `(impl->m_num_layers * 2) + 3` is used in multiple places, what about factoring it into a method (of `Impl` for instance) and using it instead?

It would also make sense to explain the expression (why the multiplication by 2? Why adding 3).

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,908 @@
+                m3hash);
+        }
+
+    private:
```
Indent with 2 spaces.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,908 @@
+    }
+
+    string manifest_str;
+    for (auto& hash : name_map)
```
Make const.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,908 @@
+        sprintf(hash_hex, "%08x", hash.first);
+        manifest_str.append(format("\"{0}\":\"{1}\",", hash.second, string(hash_hex)));
+    }
+    if (manifest_str.size() > 0)
```
`!empty()`

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,908 @@
+    {
+        char hash_hex[9];
+        sprintf(hash_hex, "%08x", hash.first);
+        manifest_str.append(format("\"{0}\":\"{1}\",", hash.second, string(hash_hex)));
```
This is fairly inefficient, use `std::stringstream` instead.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,908 @@
+    if (manifest_str.size() > 0)
+        manifest_str.pop_back();
+
+    const string manifest_string = format("{{0}}", manifest_str);
```
Farily inefficient too, I would suggest appending a `{` at the beginning and a `}` at the end.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,908 @@
+                    weight_map.insert(m3hash, old_value + weight); // possible race conditions writing to pixels on the tile edges
+                }
+            }
+            return;
```
Remove.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,908 @@
+                    const float weight = m_renderer_filter->evaluate(rx - dx, ry - dy);
+                    WeightMap& weight_map = m_pixel_samples[(m_tile_origin_y + ry) * m_frame_width + (m_tile_origin_x + rx)];
+                    const float old_value = weight_map.get(m3hash);
+                    weight_map.insert(m3hash, old_value + weight); // possible race conditions writing to pixels on the tile edges
```
That's a big problem!

`FilteredTile` handles it by using `foundation::atomic_add()` to atomically add weights; for that to work here, `WeightMap` needs to expose a mutable refs to values.

You can look at how `FilteredTile::atomic_add()` is implemented.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,908 @@
+        map<pair<int, int>, NameMap>&   m_tile_name_map;
+        CryptomatteAOV::CryptomatteType m_layer_type;
+
+        // Copy of the FilteredTile::add().
```
Remove "the" or add "method" at the end.

Maybe "inspired" instead of "copy" since it's not actually a copy?

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,908 @@
+            int tile_pixel_x = pixel_pos.x - m_tile_origin_x;
+            int tile_pixel_y = pixel_pos.y - m_tile_origin_y;
+
+            m_tile_name_map[make_pair(m_tile_origin_x, m_tile_origin_y)][m3hash] = obj_name;    //tile_origin indexing of the map should help avoiding the race condition
```
Either it completely avoids the race condition, or it doesn't: _helping_ avoiding the race condition is not enough :)

In any case, it is completely unsafe to modify a `std::map<>` concurrently so it would first need to be populated with enough entries such that the map itself is never modified.

That said, it seems you don't even need a map at all, you just need an array indexed by tile index (i.e. tile_y * tile_count_y + tile_x).

usakhelo replied:
Great tip, thanks. I don't know why I used map of maps here.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,908 @@
+
+    image_attributes_copy.insert(layer_prefix + "/manifest", manifest_string.c_str());
+
+    vector<string> channel_names;
```
You can reserve a number of items here since you already know how many you'll push.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,908 @@
+    }
+    catch (const exception& e)
+    {
+        RENDERER_LOG_ERROR(
```
Include `renderer/global/globallogger.h`

dictoon commented:
Will need to be rebased on top of master.

usakhelo commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -243,10 +249,111 @@ namespace
+            m_map = new Entry[size];
+        }
+
+        WeightMap(const WeightMap& other)
```
I had it, then I tried to compile without it and it compiled, so I decided to remove it. Will const m_size work with assignment operator?

dictoon replied:
Ah, no, it won't, sorry. Fine as is then.

usakhelo commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -326,46 +433,64 @@ namespace
+                        if (m3hash != 0)
+                        {
+                            r = hash_to_float(m3hash);
+                            g = (static_cast<float>(m3hash << 8) / static_cast<float>(UINT32_MAX));
```
Why do I need to precompute? To multiply each time instead of dividing?

dictoon replied:
Yes, because the division is a particularly expensive operation.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,908 @@
+
+            const Vector2i pixel_pos = pixel_context.get_pixel_coords();
+            const Vector2d sample_pos = pixel_context.get_sample_position();
+            int tile_pixel_x = pixel_pos.x - m_tile_origin_x;
```
Make const.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -0,0 +1,908 @@
+
+            m_tile_name_map[make_pair(m_tile_origin_x, m_tile_origin_y)][m3hash] = obj_name;    //tile_origin indexing of the map should help avoiding the race condition
+
+            double sample_pos_x = (sample_pos.x / m_rcp_canvas_width) - pixel_pos.x;
```
Make const.

usakhelo commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -243,10 +249,111 @@ namespace
+
+        Entry* end() const
+        {
+            return &m_map[m_index];
```
I think begin/end are required for for-range loops, right?

dictoon replied:
They are, but are you using for-range loops on that data structure? Iterating over a map is not so common...

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -266,11 +266,11 @@ namespace
         {
             assert(size > 0);
-            m_map = new Entry[size];
+            m_map = new Entry[static_cast<uint32>(size)];
```
You can simply use `size` here.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -684,6 +691,8 @@ CryptomatteAOV::CryptomatteAOV(const ParamArray& params)
 
 CryptomatteAOV::~CryptomatteAOV()
 {
+    if (impl->m_tile_name_array != nullptr)
```
The check is unnecessary, it's safe to call `delete` on `nullptr`.

dictoon commented on src/appleseed/renderer/modeling/aov/cryptomatte.cpp:
```diff
@@ -639,8 +661,8 @@ namespace
                     WeightMap& weight_map = m_pixel_samples[(m_tile_origin_y + ry) * m_frame_width + (m_tile_origin_x + rx)];
-                    const float old_value = weight_map.get(m3hash);
-                    weight_map.insert(m3hash, old_value + weight); // possible race conditions writing to pixels on the tile edges
+                    float* APPLESEED_RESTRICT ptr = weight_map.get_value_ptr(m3hash);
```
`get_value_ptr()` is not thread-safe, is that OK?

usakhelo replied:
I guess it's not OK.
It's probably that it creates new entries in the dictionary, is that what makes it unsafe, right?

dictoon replied:
Yes.

usakhelo replied:
Hi @dictoon, is it possible to make thread safe dictionary without guards/mutexes? I think it should be something quick and simple right? Shouldn't block the thread I think..
Maybe it's ok to have thread collisions on some of the samples at the edge of the tiles?

usakhelo replied:
Forgot to mention - I checked arnold's cryptomatte and it looks like they don't do thread safe writes on their sample buffer. It's possible they have totally different implementation of the tiles but I was just curious.

usakhelo replied:
Hi @dictoon , I thought maybe it should have a sample storage per tile and that will make it thread safe? Sorry it takes me long time to fix properly.

usakhelo replied:
I think I misunderstood and tile doesn't write outside its boundaries, so maybe it was perfectly thread safe before.

dictoon replied:
That's correct.
